crates/rpc-types-eth/src/filter.rs

```diff
@@ -534,10 +593,163 @@ impl Filter {
+            Both(filter, log) => filter.matches(log),
+            Left(filter) => filter.is_empty(),
+            Right(_) => false,
+        })
+    }
+
+    /// Returns `true` if the filter matches the given address.
+    pub fn matches_address(&self, address: Address) -> bool {
+        self.address.matches(&address)
+    }
+
+    /// Returns `true` if the block matches the filter.
+    pub fn matches_block_range(&self, block_number: u64) -> bool {
+        let mut res = true;
+
+        if let Some(BlockNumberOrTag::Number(num)) = self.block_option.get_from_block() {
+            if *num > block_number {
+                res = false;
+            }
+        }
+
+        if let Some(to) = self.block_option.get_to_block() {
+            match to {
+                BlockNumberOrTag::Number(num) => {
+                    if *num < block_number {
+                        res = false;
+                    }
+                }
+                BlockNumberOrTag::Earliest => {
+                    res = false;
+                }
+                _ => {}
+            }
+        }
+        res
+    }
+
+    /// Returns `true` if the filter matches the given block hash.
+    pub fn matches_block_hash(&self, block_hash: B256) -> bool {
+        match self.block_option {
+            FilterBlockOption::AtBlockHash(hash) => hash == block_hash,
+            FilterBlockOption::Range { .. } => false,
+        }
+    }
+
+    /// Returns `true` if the filter matches the given block. Checks both the
+    /// block number and hash.
+    pub fn matches_block(&self, block: &BlockNumHash) -> bool {
+        self.matches_block_range(block.number) || self.matches_block_hash(block.hash)
+    }
+
+    /// Check if a [`Log`] matches the filter. This will check topics and
+    /// address.
+    ///
+    /// This checks [`Log<LogData>`], the raw, primitive type carrying un-parsed
+    /// [`LogData`].
+    ///
+    /// - For un-parsed RPC logs [`crate::Log<LogData>`], see [`Self::rpc_matches`] and
+    ///   [`Self::rpc_matches_parsed`].
+    /// - For parsed [`Log`]s (e.g. those returned by a contract), see [`Self::matches_parsed`].
+    pub fn matches(&self, log: &Log) -> bool {
+        if !self.matches_address(log.address) {
+            return false;
+        }
+
+        self.matches_topics(log.topics())
+    }
+
+    /// Check if a [`crate::Log`] matches the filter. This will check topics,
+    /// address, and block option.
+    ///
+    /// This function checks [`crate::Log<LogData>`], the RPC type carrying
+    /// un-parsed [`LogData`].
+    ///
+    /// - For parsed [`Log<T>`]s (e.g. those returned by a contract), see [`Self::matches_parsed`].
+    /// - For parsed [`crate::Log<T>`]s (e.g. those returned by a contract), see
+    ///   [`Self::rpc_matches`].
+    pub fn rpc_matches(&self, log: &crate::Log) -> bool {
+        let Some(number) = log.block_number else { return false };
+        let Some(hash) = log.block_hash else { return false };
```

Comment:
this will never match for pending logs,
should this behave like an OR

Reply:
fixed by adding `matches_log_block`